dual/dual.go

```diff
@@ -0,0 +1,280 @@
+				found[pi.ID] = struct{}{}
+				count--
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return outCh
+}
+
+// FindPeer searches for a peer with given ID
+// Note: with signed peer records, we can change this to short circuit once either DHT returns.
+func (dht *DHT) FindPeer(ctx context.Context, pid peer.ID) (peer.AddrInfo, error) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var wanInfo, lanInfo peer.AddrInfo
+	var wanErr, lanErr error
+	go func() {
+		defer wg.Done()
+		wanInfo, wanErr = dht.WAN.FindPeer(ctx, pid)
+	}()
+	go func() {
+		defer wg.Done()
+		lanInfo, lanErr = dht.LAN.FindPeer(ctx, pid)
+	}()
+
+	wg.Wait()
+
+	return peer.AddrInfo{
+		ID:    pid,
+		Addrs: append(wanInfo.Addrs, lanInfo.Addrs...),
+	}, mergeErrors(wanErr, lanErr)
+}
+
+func mergeErrors(a, b error) error {
+	if a == nil && b == nil {
+		return nil
+	} else if a != nil && b != nil {
+		return fmt.Errorf("%v, %v", a, b)
+	} else if a != nil {
+		return a
+	}
+	return b
+}
+
+// Bootstrap allows callers to hint to the routing system to get into a
+// Boostrapped state and remain there.
+func (dht *DHT) Bootstrap(ctx context.Context) error {
+	erra := dht.WAN.Bootstrap(ctx)
+	errb := dht.LAN.Bootstrap(ctx)
+	return mergeErrors(erra, errb)
+}
+
+// PutValue adds value corresponding to given Key.
+func (dht *DHT) PutValue(ctx context.Context, key string, val []byte, opts ...routing.Option) error {
+	if dht.activeWAN() {
+		return dht.WAN.PutValue(ctx, key, val, opts...)
+	}
+	return dht.LAN.PutValue(ctx, key, val, opts...)
+}
+
+// GetValue searches for the value corresponding to given Key.
+func (d *DHT) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
+	reqCtx, cncl := context.WithCancel(ctx)
+	defer cncl()
+
+	resChan := make(chan []byte)
+	defer close(resChan)
+	errChan := make(chan error)
+	defer close(errChan)
+	runner := func(impl *dht.IpfsDHT, valCh chan []byte, errCh chan error) {
+		val, err := impl.GetValue(reqCtx, key, opts...)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		valCh <- val
+	}
+	go runner(d.WAN, resChan, errChan)
+	go runner(d.LAN, resChan, errChan)
+
+	var err error
+	var val []byte
+	select {
+	case val = <-resChan:
+		cncl()
+	case err = <-errChan:
+	}
+
+	// Drain or wait for the slower runner
+	select {
+	case secondVal := <-resChan:
+		if val == nil {
+			val = secondVal
+		}
+	case secondErr := <-errChan:
+		if err != nil {
+			err = mergeErrors(err, secondErr)
+		} else if val == nil {
+			err = secondErr
+		}
+	}
+	return val, err
+}
+
+// SearchValue searches for better values from this value
+func (dht *DHT) SearchValue(ctx context.Context, key string, opts ...routing.Option) (<-chan []byte, error) {
+	p := helper.Parallel{Routers: []routing.Routing{dht.WAN, dht.LAN}, Validator: dht.WAN.Validator}
+	return p.SearchValue(ctx, key, opts...)
+}
+
+// GetPublicKey returns the public key for the given peer.
+func (d *DHT) GetPublicKey(ctx context.Context, pid peer.ID) (ci.PubKey, error) {
```

Comment:
Unlike GetValue, in _this_ case, we should take the first one we find as all keys are equivalent.